Packages/com.unity.inputsystem/InputSystem/Actions/InputActionParameters.cs

```diff
@@ -0,0 +1,805 @@
+            public FieldInfo field;
+            public int bindingIndex;
+        }
+
+        // Finds all instances of a parameter in one or more actions.
+        private struct ParameterEnumerable : IEnumerable<Parameter>
+        {
+            private InputActionState m_State;
+            private ParameterOverride m_Parameter;
+            private int m_MapIndex;
+
+            public ParameterEnumerable(InputActionState state, ParameterOverride parameter, int mapIndex = -1)
+            {
+                m_State = state;
+                m_Parameter = parameter;
+                m_MapIndex = mapIndex;
+            }
+
+            public ParameterEnumerator GetEnumerator()
+            {
+                return new ParameterEnumerator(m_State, m_Parameter, m_MapIndex);
+            }
+
+            IEnumerator<Parameter> IEnumerable<Parameter>.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+        }
+
+        private struct ParameterEnumerator : IEnumerator<Parameter>
+        {
+            private InputActionState m_State;
+            private int m_MapIndex;
+            private int m_BindingCurrentIndex;
+            private int m_BindingEndIndex;
+            private int m_InteractionCurrentIndex;
+            private int m_InteractionEndIndex;
+            private int m_ProcessorCurrentIndex;
+            private int m_ProcessorEndIndex;
+
+            private InputBinding m_BindingMask;
+            private Type m_ObjectType;
+            private string m_ParameterName;
+            private bool m_MayBeInteraction;
+            private bool m_MayBeProcessor;
+            private bool m_MayBeComposite;
+            private bool m_CurrentBindingIsComposite;
+            private object m_CurrentObject;
+            private FieldInfo m_CurrentParameter;
+
+            public ParameterEnumerator(InputActionState state, ParameterOverride parameter, int mapIndex = -1)
+                : this()
+            {
+                m_State = state;
+                m_ParameterName = parameter.parameter;
+                m_MapIndex = mapIndex;
+                m_ObjectType = parameter.objectType;
+                m_MayBeComposite = m_ObjectType == null || typeof(InputBindingComposite).IsAssignableFrom(m_ObjectType);
+                m_MayBeProcessor = m_ObjectType == null || typeof(InputProcessor).IsAssignableFrom(m_ObjectType);
+                m_MayBeInteraction = m_ObjectType == null || typeof(IInputInteraction).IsAssignableFrom(m_ObjectType);
+                m_BindingMask = parameter.bindingMask;
+                Reset();
+            }
+
+            private bool MoveToNextBinding()
+            {
+                // Find a binding that matches our mask.
+                while (true)
+                {
+                    ++m_BindingCurrentIndex;
```

Comment:
minor: but why not setting m_BindingCurrentIndex to zero in Reset() and simplify this to e.g. 
`for(; m_BindingCurrentIndex != m_BindingEndIndex; ++m_BindingCurrentIndex)` - e.g. get rid of while true, the first paragraph here? Wonder whether "// Minus one to account for first MoveNext()." would be needed for enumeration pattern here?

Reply:
That doesn't work correctly. The iterator must have an initial state of "one before first element" and `MoveNext()` must make progress. The suggested for loop and similar patterns won't.